app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -232,53 +250,73 @@ private void handleLastDataPointMessage(final LastDataPointMessage message) {
             _metrics.recordCounter("rollup/generator/last_data_point_message/success", 0);
             LOGGER.warn()
                     .setMessage("Failed to get last data point for metric.")
-                    .addData("metricName", metricName)
+                    .addData("sourceMetricName", sourceMetricName)
+                    .addData("rollupMetricName", rollupMetricName)
                     .setThrowable(throwable)
                     .log();
 
             getSelf().tell(
                     new FinishRollupMessage.Builder()
-                            .setMetricName(metricName)
+                            .setMetricName(sourceMetricName)
                             .setPeriod(period)
                             .setFailure(throwable)
                             .build(),
                     ActorRef.noSender());
         } else {
             _metrics.recordCounter("rollup/generator/last_data_point_message/success", 1);
-            final Instant startOfRecentClosedPeriod = period.recentStartTime(_clock.instant());
-            final Instant lastDataPoint = message.getLastDataPointTime().orElse(Instant.EPOCH);
+
+            // Example:
+            //
+            // Consider a minutely metric that has just hit 00:00 UTC 3 Jan
+            //
+            //                 22:00         23:00         00:00      startOfLastEligiblePeriod
+            //                   |             |             |
+            // minutely  x x x x x x x x x x x x x x x x x x x                  N/A
+            //   hourly          x             x             |        23:00 2 Jan (1 period  ago)
+            //    daily          |             |             |        00:00 1 Jan (2 periods ago)
+            //
+            // Hourly pulls from the minutely, and sees that the most recently closed period
+            // is before the most recent datapoint in minutely, and so we can roll up.
+            //
+            // Daily pulls from hourly, but the latest hourly datapoint is at 23:00 < 00:00, the end
+            // of the most recent period. Therefore we cannot roll-up just yet.
+
+            final Instant lastRollupDataPoint = message.getRollupLastDataPointTime().orElse(Instant.EPOCH);
+
+            final Instant startOfLastEligiblePeriod =
+                message.getSourceLastDataPointTime()
+                        .map(period::recentStartTime)
+                        .orElse(Instant.EPOCH);
 
             // If the most recent period aligned start time is after the most recent datapoint then
             // we need to run the rollup, otherwise we can skip this and just send a finish message.
-            if (startOfRecentClosedPeriod.isAfter(lastDataPoint)) {
-                final String rollupMetricName = metricName + period.getSuffix();
+            if (startOfLastEligiblePeriod.isAfter(lastRollupDataPoint)) {
```

Comment:
(Does the hourly rollup for 10:00-11:00 land on 10:00 or 11:00? There's opportunity for a fencepost error here.)

Reply:
10:00 - we align with the start time. So in this case, if the start of the next eligible period is after the last rollup time, that means this period is unaccounted for and can be rolled up.